### PR TITLE
Upgrade to Lucene 6.5.0

### PR DESCRIPTION
**Pull request by jimczi**
The last commit is worth reviewing. It contains a small hack to retrieve the merge rate limiter hidden by https://issues.apache.org/jira/browse/LUCENE-7700.

### CODE REVIEW

File: core/src/main/java/org/apache/lucene/index/OneMergeHelper.java [hunk 1]
```diff
@@ -27,4 +29,31 @@ private OneMergeHelper() {}
...
     }
+
+    /**
+     * The current MB per second rate limit for this merge.
+     **/
+    public static double getMbPerSec(Thread thread, MergePolicy.OneMerge merge) {
+        if (thread instanceof ConcurrentMergeScheduler.MergeThread) {
+            return ((ConcurrentMergeScheduler.MergeThread) thread).rateLimiter.getMBPerSec();
+        }
+        return Double.POSITIVE_INFINITY;
```
Comment (jpountz) [inline core/src/main/java/org/apache/lucene/index/OneMergeHelper.java]: should we `assert false` here to get test failures if the thread is not a merge thread?

File: core/src/main/java/org/apache/lucene/index/OneMergeHelper.java [hunk 2]
```diff
@@ -27,4 +29,31 @@ private OneMergeHelper() {}
...
+    public static long getTotalBytesWritten(Thread thread,
+                                            MergePolicy.OneMerge merge) throws IOException {
+        /**
+         * TODO: The number of bytes written during the merge should be accessible in OneMerge.
+         */
+        if (thread instanceof ConcurrentMergeScheduler.MergeThread) {
+            return ((ConcurrentMergeScheduler.MergeThread) thread).rateLimiter
+                .getTotalBytesWritten();
+        }
+        return merge.totalBytesSize();
```
Comment (jpountz) [inline core/src/main/java/org/apache/lucene/index/OneMergeHelper.java]: same here?

**Review by jpountz — APPROVED** [review]
I think the hack is fine, especially given that the information is only used to write a log message.

File: core/src/main/java/org/apache/lucene/index/OneMergeHelper.java (hunk 1 above)
Comment (jimczi) [inline core/src/main/java/org/apache/lucene/index/OneMergeHelper.java]: Good idea, I'll add it.

**Pull request merged**